File: dotnet/source/amp.eventing/streaming/DefaultStreamingBus.cs
using amp.messaging;
using cmf.bus;
using cmf.eventing.patterns.streaming;
using Common.Logging;
using System;
using System.Collections.Generic;

namespace amp.eventing.streaming
{
    public class DefaultStreamingBus : DefaultEventBus, IStandardStreamingEventBus
    {
        protected ILog Log;
        
        private readonly IEnvelopeBus _envelopeBus;
        private readonly IEventStreamFactory _eventStreamFactory;
        private readonly Dictionary<string, IEventStream> _eventStreams;
        
        protected int _batchLimit = 10;

        public DefaultStreamingBus(IEnvelopeBus envelopeBus
            , List<IMessageProcessor> inboundChain
            , List<IMessageProcessor> outboundChain)
            : this (envelopeBus,  new DefaultEventStreamFactory(), inboundChain, outboundChain)
        {
            _eventStreamFactory.EventBus = this;
        }

        public DefaultStreamingBus(IEnvelopeBus envelopeBus
            , IEventStreamFactory eventStreamFactory
            , List<IMessageProcessor> inboundChain
            , List<IMessageProcessor> outboundChain)
            : base(envelopeBus, inboundChain, outboundChain)
        {
            Log = LogManager.GetLogger(this.GetType());

            _envelopeBus = envelopeBus;
            
            _eventStreamFactory = eventStreamFactory;
            _eventStreams = new Dictionary<string, IEventStream>();
           
        }

        public IEventStream CreateStream(string topic)
        {
            if (false == _eventStreams.ContainsKey(topic))
            {
                _eventStreamFactory.Topic = topic;
                IEventStream eventStream = _eventStreamFactory.GenerateEventStream();
                _eventStreams.Add(topic, eventStream);
                return eventStream;
            }
            else
            {
                return _eventStreams[topic];
            }
        }

        public void RemoveStream(string topic) 
        {
            if (_eventStreams.ContainsKey(topic))
            {
                _eventStreams.Remove(topic);
            }
        }

        public void PublishChunkedSequence<TEvent>(ICollection<TEvent> dataSet)
        {
            Log.Debug("enter publish to chunked sequence");
            IEventStream eventStream = null;
            string topic = null;

            ValidateEventCollection(dataSet);
            
            try
            {
                foreach (TEvent eventItem in dataSet)
                {
                    if (null == eventStream)
                    {
                        topic = eventItem.GetType().ToString();
                        eventStream = new DefaultEventStream(this, topic); //Skipping use of the factory so that we ensure sequencing based event stream is used
                        eventStream.BatchLimit = _batchLimit;
                        //Notify the receiver what the size of the collection will be
                        Publish(new CollectionSizeNotifier(dataSet.Count, topic, eventStream.SequenceId));
                        _eventStreams.Add(topic, eventStream);
                    }

                    eventStream.Publish(eventItem);
                }
            }
            finally
            {
                if (null != eventStream)
                {
                    eventStream.Dispose();
                    RemoveStream(topic);
                }
            }
            Log.Debug("leave publish to chunked sequence");
        }

        private void ValidateEventCollection<TEvent>(ICollection<TEvent> eventEnumerator)
        {
            if (null == eventEnumerator)
            {
                throw new ArgumentException("Cannot publish from a null event iterator.");
            }
        }

        public void SubscribeToCollection<TEvent>(IStreamingCollectionHandler<TEvent> handler)
        {
            Log.Debug("enter SubscribeToCollection");
            StreamingCollectionRegistration<TEvent> registration = new StreamingCollectionRegistration<TEvent>(handler, this.ProcessInboundCallback);
            _envelopeBus.Register(registration);
            Log.Debug("leave SubscribetoCollection");
        }

        public void SubscribeToReader<TEvent>(IStreamingReaderHandler<TEvent> handler)
        {
            Log.Debug("enter SubscribeToReader");
            StreamingReaderRegistration<TEvent> registration = new StreamingReaderRegistration<TEvent>(handler, this.ProcessInboundCallback);
            _envelopeBus.Register(registration);
            Log.Debug("leave SubscribeToReader");
        }

        public IEnvelopeBus EnvelopeBus
        {
            get
            {
                return _envelopeBus;
            }
        }

        public object ProcessInboundCallback(Envelope env)
        {
            Log.Debug("Enter InterceptEvent");

            var context = new MessageContext(MessageContext.Directions.In, env);

            this.ProcessMessage(context, () =>
            {
                Log.Info("Completed inbound processing - invoking IEventHandler");

            });

            Log.Debug("Leave InterceptEvent");
            return context.Message;
        }

        public void ProcessMessage(MessageContext context, Action continueProcessing)
        {
            _eventProducer.ProcessMessage(context, continueProcessing);
        }
    }
}

File: java/amp-eventing/src/main/java/amp/eventing/streaming/DefaultStreamingBus.java
package amp.eventing.streaming;


import amp.eventing.*;
import cmf.bus.IEnvelopeBus;
import cmf.eventing.IEventFilterPredicate;
import cmf.eventing.IEventHandler;
import cmf.eventing.patterns.streaming.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

public class DefaultStreamingBus extends DefaultEventBus implements IStandardStreamingEventBus, IInboundProcessorCallback {
    protected static final Logger log = LoggerFactory.getLogger(DefaultStreamingBus.class);
    private IEventStreamFactory eventStreamFactory;
    private Map<String, IEventStream> eventStreams;
    /**
     * Default setting for {@link cmf.eventing.patterns.streaming.IStreamingEventBus} to stream events in batches.
     * Allows for tuning by setting this to a different value based on size of events.
     */
    protected int batchLimit = 10;

    public DefaultStreamingBus(IEnvelopeBus envelopeBus) {
        super(envelopeBus);
        initializeDefaults();
    }

    public DefaultStreamingBus(IEnvelopeBus envelopeBus, List<IEventProcessor> inboundProcessors,
                               List<IEventProcessor> outboundProcessors) {
        super(envelopeBus, inboundProcessors, outboundProcessors);
        initializeDefaults();
    }


    public DefaultStreamingBus(IEnvelopeBus envelopeBus, List<IEventProcessor> inboundProcessors,
                               List<IEventProcessor> outboundProcessors,
                               IEventStreamFactory eventStreamFactory) {
        super(envelopeBus, inboundProcessors, outboundProcessors);
        this.eventStreamFactory = eventStreamFactory;
        this.eventStreams = new HashMap<String, IEventStream>();
    }

    private void initializeDefaults() {
        this.eventStreamFactory = new DefaultEventStreamFactory();
        this.eventStreamFactory.setEventBus(this);
        this.eventStreams = new HashMap<String, IEventStream>();
    }

    @Override
    public IEventStream createStream(String topic) {
        if (false == this.eventStreams.containsKey(topic)) {
            eventStreamFactory.setTopic(topic);
            IEventStream eventStream = eventStreamFactory.generateEventStream();
            this.eventStreams.put(topic, eventStream);
            return eventStream;
        } else {
            return this.eventStreams.get(topic);
        }

    }

    public void removeStream(String topic) {
        if (this.eventStreams.containsKey(topic)) {
            this.eventStreams.remove(topic);
        }
    }

    @Override
    public <TEVENT> void publishChunkedSequence(Collection<TEVENT> dataSet) throws Exception {
        log.debug("enter publish to chunked sequence");
        IEventStream eventStream = null;
        String topic = null;

        validateEventCollection(dataSet);

        try {

            for (TEVENT eventItem : dataSet) {
                if (null == eventStream) {
                    topic = eventItem.getClass().getCanonicalName();
                    eventStream = new DefaultEventStream(this, topic); //Skipping use of the factory so that we ensure sequencing based event stream is used
                    eventStream.setBatchLimit(this.batchLimit);
                    //Notify the receiver what the size of the collection will be
                    this.publish(new CollectionSizeNotifier(dataSet.size(), topic, eventStream.getSequenceId()));
                    eventStreams.put(topic, eventStream);
                }

                eventStream.publish(eventItem);
            }
        } finally {
            if (null != eventStream) {
                eventStream.dispose();
                removeStream(topic);
            }
        }
        log.debug("leave publish to chunked sequence");
    }


    private <TEVENT> void validateEventCollection(Collection<TEVENT> eventIterator) throws Exception {
        if (null == eventIterator) {
            throw new IllegalArgumentException("Cannot publish a null event iterator");
        }
    }

    @Override
    public <TEVENT> void subscribeToCollection(IStreamingCollectionHandler<TEVENT> handler) throws Exception {
        log.debug("enter subscribeToCollection");
        StreamingCollectionRegistration<TEVENT> registration = new StreamingCollectionRegistration<TEVENT>(handler, this);
        envelopeBus.register(registration);
        log.debug("leave subscribeToCollection");
    }

    @Override
    public <TEVENT> void subscribeToReader(IStreamingReaderHandler<TEVENT> handler) throws Exception {
        log.debug("enter subscribeToReader");
        StreamingReaderRegistration<TEVENT> registration = new StreamingReaderRegistration<TEVENT>(handler, this);
        envelopeBus.register(registration);
        log.debug("leave subscribeToReader");
    }

    @Override
    public <TEVENT> void subscribe(final IEventHandler<TEVENT> eventHandler, final IEventFilterPredicate filterPredicate)
            throws Exception {
        log.debug("enter default streaming bus subscribe.");

    }

    @Override
    public List<IEventProcessor> getInboundProcessors() {
        return this.inboundProcessors;
    }

    @Override
    public List<IEventProcessor> getOutboundProcessors() {
        return this.outboundProcessors;
    }

    @Override
    public IEnvelopeBus getEnvelopeBus() {
        return this.envelopeBus;
    }

}
